### PR TITLE
remove node-grok from the default juttle installation

### LINKED ISSUE #289: make node-grok optional when installing
there's no reason to depend on the node-grok library that pulls in a few other things (mainly the oniguruma parser) and makes our install a good bit longer and depending on some c code compiling which we can avoid and simply let users konw to `npm install xyz` if they intend on taking advantage of the grok parser.

### PR DESCRIPTION
**Pull request by rlgomes**
fixes #289

now node-grok will not be installed by npm and the user that has any
interest in using the grok parser will be prompted to install it
themselves (currently off our own fork).

### CODE REVIEW

**Comment by rlgomes** [comment]
@demmer made the magic happen, unfortunately i can't find a good way to test this than what I did locally to verify you can use `read file` without having the `node-grok` packaged installed and things work till you attempt to use `-format 'grok'` and then are told to:

```
rlgomes@x230> ./bin/juttle -e "read file -file '/var/log/syslog' -format 'grok' -pattern '%{SYSLOGLINE}'"
Error: Grok parser requires a manual installation of node-grok like so:
 > npm install rlgomes/node-grok#b3b810f4277d7b9d765853214bf20adfdb010ebe
```

Let me know what you think I can also add a similar set of instructions on installing node-grok to the docs.

File: lib/adapters/parsers/index.js [hunk 1]
```diff
@@ -28,13 +27,6 @@ module.exports = {
...
 
-        if (parser) {
-            return new parser(options);
-        } else {
-            throw errors.runtimeError('RT-INVALID-OPTION-VALUE',{
-                option: 'format',
-                supported: _.keys(parserLookup).join(', ')
-            });
-        }
+        return new parser(options);
```
Comment (demmer) [inline lib/adapters/parsers/index.js]: It seems to be simpler to keep this file the way it was and just throw an error in the grok parser initialize() function if you can't require grok.

Comment (rlgomes) [inline lib/adapters/parsers/index.js]: Sure

**Comment by rlgomes** [comment]
@demmer a lot less lines changed, this does make running the tests locally as a developer a bit of a pain when it fails to tell you to run the `npm install node-grok` command.

**Comment by demmer** [comment]
I like this implementation better but I don't like the idea that a run of `gulp test` will fail unless we install `node-grok` by hand, since that kinda defeats the point of making this optional.

How about if we add a little test utility that checks if `node-grok` is installed and then we use that instead of `it()`. Something like:

```
// grok-test-utils: simple wrapper to automatically skip grok tests if node-grok isn't installed

var grok;
try {
    grok = require('node-grok');
} catch(err) {
    console.log('WARNING: skipping all tests that depend on grok');
}
function ifGrok(description, function) {
    if (grok) {
        it(description, function);
    } else{
        it.skip(description, function);
    }
}
```

Then all the tests that depend on grok would use `ifGrok(...)` instead of `it(...)`.

**Comment by rlgomes** [comment]
@demmer checkout the last commit I built a test it that can accept an options argument and then you can do special requirements. This can be used or our case here of needed a specific module that we don't want to install by default but can also be used for other requirements like depednent services, etc. Let me know what you think.

**Comment by demmer** [comment]
LGTM. +1

**Comment by dmajda** [comment]
Late to the party, but I don’t think the module check should be conflated with the `it` function — they are two different things. I’d find it much cleaner to see something like this in the tests:

```javascript
ifModuleInstalled('node-grok', function() {
    it('can instantiate a grok parser', function() {
        var parser = parsers.getParser('grok');
        expect(parser).to.not.be.undefined();
    });

    // ...
});
```

Or perhaps more generically:

```javascript
condition({ installedModules: 'node-grok' }, function() {
    it('can instantiate a grok parser', function() {
        var parser = parsers.getParser('grok');
        expect(parser).to.not.be.undefined();
    });

    // ...
});
```

The implementation would just pass the function with `it` blocks to `describe`/`describe.skip` with synthesized description (e.g. “with node-grok module installed”).

@rlgomes What do you think?

**Comment by rlgomes** [comment]
@dmajda not sure how would the `condition` or `ifModuleInstalled` functions modify the underlying it function calls ? so they would all show up as skipped ? instead of not appearing in the test report at all ? or just under a single skip ?

**Comment by dmajda** [comment]
@rlgomes My idea was that `ifModuleInstalled`/`condition` would just morph into `describe` or `describe.skip`, like this:

```javascript
function ifModuleInstalled(module, fn) {
    var installed = checkForModule(module);
    var description = 'with module "' + module + '" installed';
  
    (installed ? describe : describe.skip)(description, fn);
}
```

**Comment by rlgomes** [comment]
@dmajda understood the only problem is that there are a few just `it` tests in some files like these ones https://github.com/juttle/juttle/pull/290/files#diff-7a886cc67329f5d0b14d46f836d42702R197 that I'd have to move into another describe block ? This was just to remove the installing/building of the oniguruma package from building its c dependencies in our juttle runtime. I have something I'm working on which could replace the whole `node-grok` package by a much simpler depdendency that wouldn't rely on the oniguruma package.

**Comment by dmajda** [comment]
> the only problem is that there are a few just `it` tests in some files like these ones https://github.com/juttle/juttle/pull/290/files#diff-7a886cc67329f5d0b14d46f836d42702R197 that I'd have to move into another describe block ?

Not sure what you mean by “move into another describe block”. On the source code level, you’d just replace:

```javascript
withModuleIt('fails to optimized tail followed by head with -format "grok"', function() {
    //...
}, 'node-grok');

withModuleIt('can optimize nested "| head 2 | head 1" with -format "grok"', function() {
    //...
}, 'node-grok');

withModuleIt('can optimize "| head 1" with -format "grok"', function() {
    //...
}, 'node-grok');
```

with:

```javascript
ifModuleInstalled('node-grok', function() {
    it('fails to optimized tail followed by head with -format "grok"', function() {
        //...
    });

    it('can optimize nested "| head 2 | head 1" with -format "grok"', function() {
        //...
    });

    it('can optimize "| head 1" with -format "grok"', function() {
        //...
    });
});
```

I find that much cleaner.

> I have something I'm working on which could replace the whole `node-grok` package by a much simpler depdendency that wouldn't rely on the oniguruma package. 

OK, if this is temporary then it’s probably not worth tweaking.

**Pull request merged**